File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
int nn;
vector<int> adj[104];
int visited[104];
int mi = 987654321;
void init(){
    for(int i = 0; i < 104; i++){
        adj[i].clear();
    }
    fill(&visited[0], &visited[0] + 104, 0);
}

int dfs(int here){
    int cnt = 1;
    visited[here] = 1;
    for(int there: adj[here]){
        if(visited[there])continue;
         cnt += dfs(there);
    }
    return cnt;
}

void go(){
    vector<int> ret;
    for(int i = 1; i <= nn; i++){
        if(visited[i])continue;
        ret.push_back(dfs(i));
    }
    mi = min(mi, abs(ret[0] - ret[1]));
}

int solution(int n, vector<vector<int>> wires) {
    int answer = -1;
    nn = n;
    for(int i = 0; i < wires.size(); i++){
        // i는 끊어서
        init();
        for(int j = 0; j < wires.size(); j++){
            if(i == j)continue;
            int f = wires[j][0];
            int t = wires[j][1];
            adj[f].push_back(t);
            adj[t].push_back(f);
        }
        go();
    }
    answer = mi;
    return answer;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static int ans = Integer.MAX_VALUE;
    ;
    static boolean[] visited;

    public int solution(int n, int[][] wires) {
        for (int i = 0; i < wires.length; i++) {
            ArrayList<Integer>[] adj = new ArrayList[n + 1];
            for (int j = 0; j < n + 1; j++) {
                adj[j] = new ArrayList<>();
            }
            for (int j = 0; j < wires.length; j++) {
                if (i == j) continue;
                int f = wires[j][0];
                int t = wires[j][1];
                adj[f].add(t);
                adj[t].add(f);
            }
            visited = new boolean[n + 1];
            int cnt = dfs(1, adj);
            int diff = Math.abs(cnt - (n - cnt));
            ans = Math.min(ans, diff);
        }
        return ans;
    }

    static int dfs(int here, ArrayList<Integer>[] adj) {
        int ret = 1;
        visited[here] = true;
        for (int there : adj[here]) {
            if (visited[there]) continue;
            ret += dfs(there, adj);
        }
        return ret;
    }
}